cachito/workers/pkg_managers/pip.py

```diff
@@ -332,3 +335,296 @@ def _get_package_dirs(self):
+                for i, node in enumerate(field):
+                    setup_call, setup_path = self._find_setup_call(node)
+                    if setup_call is not None:
+                        setup_path.append(ASTpathelem(root_node, name, i))
+                        return setup_call, setup_path
+
+        return None, []  # No setup call under root_node
+
+    def _is_setup_call(self, node):
+        """Check if node is setup() or setuptools.setup() call."""
+        if not isinstance(node, ast.Call):
+            return False
+
+        fn = node.func
+        return (isinstance(fn, ast.Name) and fn.id == "setup") or (
+            isinstance(fn, ast.Attribute)
+            and fn.attr == "setup"
+            and isinstance(fn.value, ast.Name)
+            and fn.value.id == "setuptools"
+        )
+
+    def _get_setup_kwarg(self, arg_name):
+        """
+        Find setup() call, extract specified argument from keyword arguments.
+
+        If argument value is a variable, then what we do is only a very loose approximation
+        of how Python resolves variables. None of the following examples will work:
+
+        1) any indented blocks (unless setup() call appears under the same block)
+
+            with x:
+                name = "foo"
+
+            setup(name=name)
+
+        2) late binding
+
+            def my_setup():
+                setup(name=name)
+
+            name = "foo"
+
+            my_setup()
+
+        The rationale for not supporting these cases:
+        - it is difficult
+        - there is no use case for 1) which is both valid and possible to resolve safely
+        - 2) seems like a bad enough practice to justify ignoring it
+        """
+        if self._setup_branch is None:
+            return None
+
+        for kw in self._setup_branch.call_node.keywords:
+            if kw.arg == arg_name:
+                try:
+                    value = ast.literal_eval(kw.value)
+                    log.debug("setup kwarg %r is a literal: %r", arg_name, value)
+                    return value
+                except ValueError:
+                    pass
+
+                if isinstance(kw.value, ast.Name):
+                    log.debug("setup kwarg %r looks like a variable", arg_name)
+                    return self._get_variable(kw.value.id)
+
+                expr_type = kw.value.__class__.__name__
+                log.debug("setup kwarg %r is an unsupported expression: %s", arg_name, expr_type)
+                return None
+
+        log.debug("setup kwarg %r not found", arg_name)
+        return None
+
+    def _get_variable(self, var_name):
+        """Walk back up the AST along setup branch, look for first assignment of variable."""
+        lineno = self._setup_branch.call_node.lineno
+        node_path = self._setup_branch.node_path
+
+        log.debug("Backtracking up the AST from line %s to find variable %r", lineno, var_name)
```

Comment:
Why `var_name` is treated as printable representation of an object? (Out of curiosity: trying to understand the difference)

Reply:
I like to use repr for strings because they may fade into the surrounding text otherwise. Basically, I'd rather have
`find variable 'version'` than `find variable version`.

Another option would be to use `'%s'`, but in some cases, repr provides more readable output, so I just use `%r` everywhere for consistency.